### DISCUSSION TITLE
GitHub releases api does not show latest version.

### DISCUSSION THREAD

**Discussion by andreasgottardi** [kimai/kimai#4354] [On-Premise: Installation & Updates] [answered]
Hi,

I am using the GitHub releases API for my local updates and noticed, that version 2.0.35 is missing from https://api.github.com/repos/kimai/kimai/releases

Was it forgotten?

Kind regards,
Andreas

**Comment by kevinpapst** [comment, marked as answer]
I have no clue. You might want to ask GitHub if it should be in there. 

I am using this one for my scripts: https://api.github.com/repos/kimai/kimai/releases/latest

**Reply by andreasgottardi** [reply]
That's fine for me, thank you for the hint! Makes it even easier.

**Reply by simonschaufi** [reply]
I'm curious about your updates that you use with the GitHub API. May you want to share your script? I would be very interested in it!

**Reply by andreasgottardi** [reply]
Hi Simon,

sure thing:

```bash
#!/bin/bash

INSTALLEDVERSION=$(cat /var/www/.kimai/installedversion)
AVAILABLEVERSION=$(curl -s --location 'https://api.github.com/repos/kimai/kimai/releases/latest' | python3 -c "import sys,json;print(json.load(sys.stdin)['tag_name'])")

if [ "$INSTALLEDVERSION" = "$AVAILABLEVERSION" ]
then
        echo "Latest version ($INSTALLEDVERSION) already installed, not updating."
else
    echo "New version $AVAILABLEVERSION is available. Updating."
    su -c "cd /var/www/kimai; git fetch --tags; git checkout $AVAILABLEVERSION; composer install --optimize-autoloader -n; bin/console kimai:update" -s /bin/bash www-data
    echo "$AVAILABLEVERSION" > /var/www/.kimai/installedversion
fi
```

What you have to know is, that I store the currently installed version in `/var/www/.kimai/installedversion` where as `/var/www` the home directory of the user `www-data` is. I have to point out, that the script is not perfect. It just checks, if the installed version is not the same as the version from git. It does not check if it is actually newer but I never saw, that a version was downgraded so that's fine for me.

The following line curls the latest json from GitHub and passes it to a Python inline script which just prints the content of the tag value which then is assigned to the bash variable `AVAILABLEVERSION`.

```bash
AVAILABLEVERSION=$(curl -s --location 'https://api.github.com/repos/kimai/kimai/releases/latest' | python3 -c "import sys,json;print(json.load(sys.stdin)['tag_name'])")
```

If there is a newer version, the script executes the commands from the documentation as `www-data`:

```bash
su -c "cd /var/www/kimai; git fetch --tags; git checkout $AVAILABLEVERSION; composer install --optimize-autoloader -n; bin/console kimai:update" -s /bin/bash www-data
```

I know, a one-liner is not nice, but I had it in the history :-)

The script is basic but it works for me. It can also be executed in a cronjob for example. I have it in the path `/usr/sbin/update_kimai` and run it as root from time to time.
Hope this helps.

**Reply by simonschaufi** [reply]
Thank you very much! If you have installed `jq` (commandline JSON processor) installed, it would even be a bit easier.

**Reply by simonschaufi** [reply]
If you are curious, I have improved the script heavily and also read the actual version number from within Kimai itself. You need to have `jq` installed.

```bash
#!/usr/bin/env bash

# Required settings
kimai_dir="/var/www/kimai2/"

# https://dev.to/ifenna__/adding-colors-to-bash-scripts-48g4
RED="\e[31m"
GREEN="\e[32m"
GREEN_BOLD="\e[1;32m"
YELLOW="\e[33m"
BLUE="\e[36m"
LIGHT_GRAY="\e[37m"
END_COLOR="\e[0m"

# https://stackoverflow.com/a/37939589
function version {
  echo "$@" | awk -F. '{ printf("%d%03d%03d%03d\n", $1,$2,$3,$4); }';
}

version=$($kimai_dir/bin/console kimai:version --short)
echo -e "${BLUE}[INFO] Current version: $version${END_COLOR}";

check_url="https://api.github.com/repos/kimai/kimai/releases/latest"
echo -e "${LIGHT_GRAY}[DEBUG] Check url: ${check_url}";

response="$(curl --silent --location "${check_url}" 2>/dev/null)"
if [ -z "$response" ]; then
  echo -e "${RED}[ERROR] Could not fetch latest version from GitHub${END_COLOR}";
  exit 1;
fi

newVersion=$(jq --raw-output '.tag_name' <<< "$response")

echo -e "${BLUE}[INFO] Latest version: $newVersion${END_COLOR}";

# shellcheck disable=SC2046
# shellcheck disable=SC2086
if [ $(version $newVersion) -gt $(version $version) ]; then
  echo -e "${YELLOW}[NOTICE] New version available: $newVersion${END_COLOR}";
else
  echo -e "${GREEN_BOLD}[INFO] Version is up to date${END_COLOR}";
  exit 0;
fi

cd ${kimai_dir} || exit
git fetch --tags
git checkout $newVersion

composer install --optimize-autoloader -n
bin/console kimai:update
```

**Reply by kevinpapst** [reply]
For an already installed Kimai, use this: `bin/console kimai:version --short`

**Reply by simonschaufi** [reply]
@kevinpapst thank you for that hint, I have adjusted the script.

**Reply by andreasgottardi** [reply]
@simonschaufi Thanks a lot and sorry for my late response! I am happy to take a look at the script and implement it on my instance.

**Reply by divemasterza** [reply]
here's a small bash script I wrote a while back to update. I need to fix the ownership of the file after a git update so it has been added in the routine -> https://github.com/divemasterza/Kimai-updater
